src/server/raft_client.rs

```diff
@@ -262,32 +262,44 @@ impl Buffer for BatchMessageBuffer {
     }
 
     #[inline]
-    fn push(&mut self, msg: RaftMessage) {
-        let msg_size = Self::message_size(&msg);
+    fn push(&mut self, msg_with_time: (RaftMessage, Instant)) {
+        let msg_size = Self::message_size(&msg_with_time.0);
         // To avoid building too large batch, we limit each batch's size. Since
         // `msg_size` is estimated, `GRPC_SEND_MSG_BUF` is reserved for errors.
         if self.size > 0
             && (self.size + msg_size + self.cfg.raft_client_grpc_send_msg_buffer
                 >= self.cfg.max_grpc_send_msg_len as usize
-                || self.batch.get_msgs().len() >= self.cfg.raft_msg_max_batch_size)
+                || self.batch.len() >= self.cfg.raft_msg_max_batch_size)
         {
-            self.overflowing = Some(msg);
+            self.overflowing = Some(msg_with_time);
             return;
         }
         self.size += msg_size;
-        self.batch.mut_msgs().push(msg);
+        self.batch.push(msg_with_time);
     }
 
     #[inline]
     fn empty(&self) -> bool {
-        self.batch.get_msgs().is_empty()
+        self.batch.is_empty()
     }
 
     #[inline]
     fn flush(&mut self, sender: &mut ClientCStreamSender<BatchRaftMessage>) -> grpcio::Result<()> {
-        let batch = mem::take(&mut self.batch);
+        let mut batch_msgs = BatchRaftMessage::default();
+        self.batch.drain(..).for_each(|(msg, time)| {
```

Comment:
Consider shrinking the batch when the capacity becomes too large.

Reply:
That shouldn’t be necessary. The batch’s capacity is fixed at raft_max_batch_size. After each drain, the capacity remains the same, and it will be reused for the next push.